### PR TITLE
RUN-2950 iframe support

### PR DESCRIPTION
**Pull request by datamadic**
Makes the new frame connection behavior available behind a flag for a soft roll-out. Adapter and Runtime PRs required, I will post them shortly. 

The most important stuff to review here is the elipc_strategy and window get options changes. After that most all the changes are additions that you would have to opt into (not unimportant, but would not be seen by someone, say, rolling forward to this version)  

See the [iframe proposal wiki](https://github.com/openfin/Internal-Wiki/issues/13) for a full up to date description. 

[tests well](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59d5b2a27cb79f732d10c44b)

### CODE REVIEW

File: src/shapes.ts [hunk 1]
```diff
@@ -20,6 +20,28 @@ export interface Identity {
...
+}
+
+export interface ChildFrameInfo extends FrameInfo {
+    frameRoutingId: number;
+}
+
+/**
+ * Window bounds
+ */
+export interface WindowBounds {
```
Comment (whyn07m3) [inline src/shapes.ts]: We have `Rectangle` in Electron

File: src/browser/core_state.ts [hunk 2]
```diff
@@ -401,15 +399,15 @@ export function getMainWindowOptions(id: number): Shapes.WindowOptions|void {
...
     }
 
     if (!app.appObj) {
         console.warn('getMainWindowOptions - app opts not found', arguments);
-        return; //throw new Error('getMainWindowOptions - app opts not found');
+        return;
     }
 
-    // console.log('getMainWindowOptions', app.appObj._options);
+
```
Comment (whyn07m3) [inline src/browser/core_state.ts]: "...Really makes the code breathe..." ;)

File: src/browser/api_protocol/transport_strategy/elipc_strategy.ts [hunk 3]
```diff
@@ -61,10 +69,30 @@ export class ElipcStrategy extends ApiTransportBase<MessagePackage> {
         electronIpc.ipc.on(electronIpc.channels.WINDOW_MESSAGE, this.onMessage.bind(this));
     }
 
-    public send(identity: any, payload: any): void {
-        const window = coreState.getWindowByUuidName(identity.uuid, identity.name);
-        if (window && !window.browserWindow.isDestroyed()) {
-            window.browserWindow.send(electronIpc.channels.CORE_MESSAGE, JSON.stringify(payload));
+    public send(identity: any, payloadObj: any): void {
```
Comment (whyn07m3) [inline src/browser/api_protocol/transport_strategy/elipc_strategy.ts]: Let's have typed `identity` here?

File: src/browser/api_protocol/index.ts [hunk 4]
```diff
@@ -22,6 +22,7 @@ import {
     registerMiddleware as registerExternalConnAuthMiddleware
 } from './api_handlers/authorization';
 import { init as initClipboardAPIHandler } from './api_handlers/clipboard';
+const FrameApiHandler = require('./api_handlers/frame').FrameApiHandler;
```
Comment (whyn07m3) [inline src/browser/api_protocol/index.ts]: Let's `import` here

File: src/browser/api_protocol/api_handlers/window.js [hunk 5]
```diff
@@ -320,6 +321,14 @@ function hideWindow(identity, message, ack) {
     ack(successAck);
 }
 
+
+function getAllFrames(identity, message) {
+    const payload = message.payload;
```
Comment (whyn07m3) [inline src/browser/api_protocol/api_handlers/window.js]: `const { payload } = message;`

File: src/browser/api_protocol/api_handlers/frame.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import * as apiProtocolBase from './api_protocol_base';
+import { Frame } from '../../api/frame';
+import { Identity } from '../../../shapes';
+import { ActionSpecMap } from '../shapes';
```
Comment (whyn07m3) [inline src/browser/api_protocol/api_handlers/frame.ts]: Would be nice to see it alphabetized

File: src/browser/api_protocol/api_handlers/frame.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+const successAck: object = {
+    success: true
+};
+
+export class FrameApiHandler {
```
Comment (whyn07m3) [inline src/browser/api_protocol/api_handlers/frame.ts]: Some other API handlers has migrated (forward) to _functional_ style, e.g. _window api handler_ (`init` function)

File: src/browser/api_protocol/api_handlers/frame.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        'get-frame-info': this.getInfo,
+        'get-parent-window': this.getParentWindow
+    };
+
+    constructor() {
+        apiProtocolBase.registerActionMap(this.actionMap);
+    }
+
+    private getInfo(identity: Identity, message: any) {
+        const frameIdentity: any = apiProtocolBase.getTargetWindowIdentity(message.payload);
```
Comment (whyn07m3) [inline src/browser/api_protocol/api_handlers/frame.ts]: Can it be typed here?

Comment (datamadic) [reply src/browser/api_protocol/api_handlers/frame.ts]: 👍 will update

File: src/browser/api_protocol/api_handlers/frame.ts (hunk 6 above)
Comment (datamadic) [reply src/browser/api_protocol/api_handlers/frame.ts]: 👍 will update

File: src/browser/api_protocol/index.ts (hunk 4 above)
Comment (datamadic) [reply src/browser/api_protocol/index.ts]: 👍 will update

File: src/browser/api_protocol/api_handlers/window.js (hunk 5 above)
Comment (datamadic) [reply src/browser/api_protocol/api_handlers/window.js]: 👍 will update

File: src/browser/api_protocol/transport_strategy/elipc_strategy.ts (hunk 3 above)
Comment (datamadic) [reply src/browser/api_protocol/transport_strategy/elipc_strategy.ts]: 👍 will update

File: src/browser/api/frame.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import ofEvents from '../of_events';
+import { Identity } from '../../shapes';
+import route from '../../common/route';
+const coreState = require('../core_state');
```
[the remark this answers is not in the document]
Comment (datamadic) [reply src/browser/api/frame.ts]: 😆  👍 will update

File: src/shapes.ts (hunk 1 above)
Comment (datamadic) [reply src/shapes.ts]: 👍 will update

File: src/browser/core_state.ts (hunk 2 above)
Comment (datamadic) [reply src/browser/core_state.ts]: 👍 will update

File: src/browser/api_protocol/api_handlers/frame.ts (hunk 7 above)
Comment (datamadic) [reply src/browser/api_protocol/api_handlers/frame.ts]: This one I'd rather leave for now, it is a class, so just the `new` feels natural enough for now and I'd rather not refactor atm

File: src/browser/api/window.js [hunk 10]
```diff
@@ -1072,11 +1116,13 @@ Window.getNativeId = function(identity) {
...
 
-
 Window.getOptions = function(identity) {
-    let browserWindow = getElectronBrowserWindow(identity, 'get options for');
 
-    return browserWindow._options;
+    try {
+        return getElectronBrowserWindow(identity, 'get options for')._options;
+    } catch (e) {
+        return Object.assign({ entityType: 'iframe' }, identity);
```
Comment (HarsimranSingh) [inline src/browser/api/window.js]: I don't think `getOptions` should have these disparate capabilities

File: src/browser/api_protocol/api_handlers/event_listener.js [hunk 11]
```diff
@@ -87,6 +85,42 @@ function EventListenerApiHandler() {
...
+            subscribe: function(identity, type, payload, cb) {
+                const {
+                    uuid,
+                    name
+                } = payload;
+                const frameIdentity = apiProtocolBase.getTargetWindowIdentity(payload);
+                const targetUuid = frameIdentity.uuid;
+                const islocalWindow = !!coreState.getWindowByUuidName(targetUuid, targetUuid);
+                const localUnsub = Frame.addEventListener(identity, frameIdentity, type, cb);
+                let remoteUnSub;
```
Comment (HarsimranSingh) [inline src/browser/api_protocol/api_handlers/event_listener.js]: Can this be moved down closer to the conditional where it's assigned?

File: src/browser/api_protocol/api_handlers/system.js [hunk 12]
```diff
@@ -252,6 +253,12 @@ function SystemApiHandler() {
         ack(dataAck);
     }
 
+    function getEntityInfo(identity, message, ack, nack) {
+        const { uuid, name } = message.payload;
+
+        return System.getEntityInfo({ uuid, name });
```
Comment (HarsimranSingh) [inline src/browser/api_protocol/api_handlers/system.js]: ack or nack?

File: src/browser/api_protocol/api_handlers/window.js [hunk 13]
```diff
@@ -320,6 +321,14 @@ function hideWindow(identity, message, ack) {
...
+    const windowIdentity = apiProtocolBase.getTargetWindowIdentity(payload);
+
+    return Window.getAllFrames(windowIdentity);
```
Comment (HarsimranSingh) [inline src/browser/api_protocol/api_handlers/window.js]: ack or nack?

**Comment by datamadic** [comment]
[updated tests](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59dd20997cb79f732d10c487)

File: src/browser/api_protocol/api_handlers/event_listener.js (hunk 11 above)
Comment (datamadic) [reply src/browser/api_protocol/api_handlers/event_listener.js]: yeah, but this is basically a copy/past of the other subscription providers around it. I think they just all be "fixed" en mass (made a factory function or the like) at some other time

File: src/browser/api_protocol/api_handlers/system.js (hunk 12 above)
Comment (datamadic) [reply src/browser/api_protocol/api_handlers/system.js]: just the return is enough here, it will auto ack, or nack if it throws

**Comment by datamadic** [comment]
[updated tests](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59de3ba4b640c27247e63694)

**Pull request merged**